client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,380 @@
+    });
+  }
+
+  setTabLabelIndex() {
+    (this.labelTargets as TabLink[]).forEach((label, index) => {
+      // eslint-disable-next-line no-param-reassign
+      label.index = index;
+    });
+  }
+
+  setURLHash(tabId: string) {
+    if (!window.history.state || window.history.state.tabContent !== tabId) {
+      // Add a new history item to the stack
+      window.history.pushState({ tabContent: tabId }, '', `#${tabId}`);
+    }
+  }
+
+  setTabByURLHash() {
+    if (window.location.hash) {
+      const cleanedHash = window.location.hash
+        .replace(/[^\w\-#]/g, '')
+        .replace('#', '');
+      if (cleanedHash) {
+        this.selectedValue = cleanedHash;
+      } else {
+        // The hash doesn't match a tab on the page then select first tab
+        this.selectFirstTab();
+      }
+    }
+  }
+
+  selectFirstTab() {
+    const href = this.labelTargets[0].getAttribute('aria-controls') as string;
+    this.selectedValue = href;
+  }
+
+  animateIn(tabContent: HTMLElement) {
+    const selectedClasses = this.selectedClasses;
+    afterTransition(
+      tabContent,
+      // If there are no classes to add, we can skip the delay before hiding.
+      selectedClasses.length
+        ? { maxDelay: this.transitionValue }
+        : { maxDelay: 0 },
+    )(tabContent.classList.add(...selectedClasses)).then(() => {
+      // eslint-disable-next-line no-param-reassign
+      tabContent.hidden = false;
+    });
+  }
+
+  animateOut(tabContent: HTMLElement) {
+    const selectedClasses = this.selectedClasses;
+    afterTransition(
+      tabContent,
+      // If there are no classes to add, we can skip the delay before hiding.
+      selectedClasses.length
+        ? { maxDelay: this.transitionValue }
+        : { maxDelay: 0 },
+    )(tabContent.classList.remove(...selectedClasses)).then(() => {
+      // eslint-disable-next-line no-param-reassign
+      tabContent.hidden = true;
+    });
+  }
+
+  hideTabContent(tabId: string) {
+    if (tabId === this.selectedValue || !this.selectedValue) {
+      return;
+    }
+
+    const tabContent = this.getTabPanelByHref(tabId);
+    if (!tabContent) {
+      return;
+    }
+    if (this.animateValue) {
+      this.animateOut(tabContent);
+    } else {
+      tabContent.hidden = true;
+    }
+
+    const tab = this.getTabLabelByHref(tabId);
+    if (!tab) {
+      return;
+    }
+    tab.setAttribute('aria-selected', 'false');
+    tab.setAttribute('tabindex', '-1');
+  }
+
+  selectNext(event: Event) {
+    const tabIndex = (event.target as IndexedEventTarget).index;
+    const tab = this.labelTargets[tabIndex + 1];
+    if (!tab) {
+      return;
+    }
+    this.selectedValue = tab.getAttribute('aria-controls') as string;
+    tab.focus();
+  }
+
+  selectPrevious(event: Event) {
+    const tabIndex = (event.target as IndexedEventTarget).index;
+    const tab = this.labelTargets[tabIndex + -1];
+    if (!tab) {
+      return;
+    }
```

Comment:
```suggestion
    if (!tab) return;
```

Minor nitpick, just reads better if it's one line.